lib/dfdutils/makedfd2vk.pl

```diff
@@ -307,55 +313,105 @@ sub checkSuffices {
         if (B.offset == 10) return VK_FORMAT_A1B5G5R5_UNORM_PACK16_KHR;
         return VK_FORMAT_B5G5R5A1_UNORM_PACK16;
       }
-    } else if (wordBytes == 4) { /* PACK32 */
+    } else if (wordBytes == 4) { /* PACK32 or 2PACK16 */
```

Comment:
2PACK16 should be wordBytes == 2, but I'm worried about the knock-on effects elsewhere.

Reply:
No. `wordBytes` will be 4. These formats are correctly, in my view, detected as packed because the offset is not a multiple of 8.

```
@param wordBytes[in,out] Pointer to a uint32_t to receive the byte size of
 *                          the channels (unpacked) or total size (packed).
 ```

Reply:
Oh, right, not the size of the "channel" word. I said I was rusty. :-) Ignore me.

Reply:
> No. wordBytes will be 4. These formats are correctly, in my view, detected as packed because the offset is not a multiple of 8. Hence wordBytes == bytesPlane0.

Well, not really, I mean they are packed, but per-channel.

This is the general mPACKn format issue we've fixed whereas typeSize is actually 2 for these formats, just like regular 16-bit channel formats.

This wordBytes value is used in legacy code as the typeSize. But the DFD does not have the expression power to determine the wordBytes/typeSize from the DFD only.

One simply cannot differentiate at the DFD level a 4-byte format with 4-byte words with two 2-byte channels packed into them (i.e. PACK32) and a 4-byte format with 2-byte words with 2-byte channels (i.e. 2PACK16), and this makes a difference from the perspective of endianness.

This is an issue that we fixed for the Vk->DFD direction through the separate Vk->typeSize conversion, but it's not resolvable for DFD->Vk unless we add a typeSize (wordBytes) input argument making it a (DFD+typeSize)->Vk conversion. Conversions between VkFormat and DFD are simply not bijective, multiple VkFormats could result in the same DFD (e.g. a hypothetical `R10X6G10X6_PACK32` format would not be differentiable from a `R10X6G10X6_2PACK16` format at the DFD level, even though they have different packing granularities and thus different in-memory representation on a LE vs BE system).

Reply:
Also, packed vs unpacked is not a correct categorization, hence the description of wordBytes itself is wrong, because mPACKn formats with m > 1 will consist of multiple, separately packed groups of channels (1 channel in each group in this case), so the whole packed vs unpacked format categorization (and the related assumptions in various parts of the code base) is wrong.

Reply:
> I could completely believe the mPACKn formats postdated this code, so the terminology could easily be misleading.

That is part of it, but also IIRC the DFD was never meant to abstract endianness, in fact it intentionally defines different representations depending on endianness (although I'm sure, @fluppeteer, you know that better), which is very different from how VkFormats are described.

Reply:
Yup, that's what I was incoherently trying to say in my first paragraph. :-)

Reply:
I've used #defines. The code is much more clear.